### PR TITLE
Add support for CNI 1.0.0 config

### PR DESCRIPTION
**Pull request by TBBle**
Trivially upgraded github.com/containernetworking/cni to v1.1.2 and expanded the list of known versions to all the ones it supports for auto-conversion.

Specifically, the internal CNI representation is still using CNI spec 0.4.0, changing that would have required deeper changes. As long as any future CNI library updates can still convert to/from 0.4.0, we should be able to trivially upgrade CNI support using the CNI library auto-conversion. It might make sense to move to 1.0.0 internally, but that isn't necessary now.

Tested with the NAT plugin and the following CNI config, mostly generated by nerdctl 1.6.2. (I had to add the DNS config myself, and the gateway and subnet generated by nerdctl were incorrect for the existing NAT network created by (I think) Docker; both are things I'm chasing up elsewhere)

```json
{
  "cniVersion": "1.0.0",
  "name": "nat",
  "nerdctlID": "d919a100ce6b45524d415d52d088d5817587c6dd8c3691b03b8063c44d043523",
  "nerdctlLabels": {
    "nerdctl/default-network": "true"
  },
  "plugins": [
    {
      "type": "nat",
      "master": "Ethernet",
      "ipam": {
        "routes": [
          {
            "gateway": "172.30.224.1"
          }
        ],
        "subnet": "172.30.224.0/20",
        "type": ""
      },
      "dns": {
        "nameservers": [ "103.167.44.10" ]
      }
    }
  ]
}
```

### CODE REVIEW

**Comment by TBBle** [comment]
Oops. I just noticed #90 which did update the internal structure to a later CNI version (and it looks like it was a pretty trivial change), but did not enable newer CNI versions to be parsed/output.

So I guess the best of both worlds would be to combine them both?

**Comment by iankingori** [comment]
Thanks for the PR @TBBle, we've been exploring the CNI 1.0.0 update for the last few weeks with @aznashwan. I've tested this locally and seems issue is we've updated  cni to `github.com/containernetworking/cni v1.1.2` but didn't push the corresponding `go mod vendor` updates for Server 2019. Server 2022 was broken before the update

**Comment by iankingori** [comment]
Thanks @TBBle for the PR, recent commit has restored CI to match state on master. Appears we oscillate between failure on 2019 and failure on 2022. I can't repro the test failure on a local Server 2019, we should perhaps consider hosting our own runner and see whether that improves  CI stability. cc: @debj1t 

@aznashwan please help review and see whether we can merge this then port your fixes over.

**Comment by aznashwan** [comment]
LGTM FWIW, thank you!

> and the gateway and subnet generated by nerdctl were incorrect for the existing NAT network created by (I think) Docker

There is indeed currently no logic in this repo which compares the network config provided to the plugins with the actual properties of the HCN network if it is pre-created. (it simply [looks the network up by name](https://github.com/microsoft/windows-container-networking/blob/b360521954ff8b432d3b5fcf76c4997d19a39b51/common/core/network.go#L404-L420) and trusts it matches)

IMO this should be fixed and the plugins should error on config mismatches with pre-existing networks, but that will be a separate patch for another day.

I'll be following up with a patch which'll add a workflow to run containerd+cri integration tests on top of both the NAT and Bridge/Overlay plugins, so I'd suggest we merge this now @debj1t

**Comment by TBBle** [comment]
> IMO this should be fixed and the plugins should error on config mismatches with pre-existing networks, but that will be a separate patch for another day.

Makes sense. I was also planning to poke at nerdctl and see if I can work out why it created a faulty CNI config; I got rabbitholed trying to make sense of the duality of HCN/HNS and forgot about that. (I _suspect without actually checking that nerdctl just output a set of hard-coded details assuming the network doesn't already exists, rather than asking HCN for details of an existing network.)

An alternative to this (which I thought I mentioned elsewhere, I think there's a discussion about CNI network name/type behaviour in another ticket) would be to actually complete the TODO in this repo and remove network-creation support, relying on the host to already have the desired network created. That would actually remove (I think...) the need to have multiple CNI network types in this codebase at all, since they all trust HCN over the user config anyway.

**Comment by aznashwan** [comment]
> I _suspect without actually checking that nerdctl just output a set of hard-coded details

You are correct, `nerdctl` will simply use [some hardcoded defaults](https://github.com/containerd/nerdctl/blob/4e3fd3b48f60fbce3b8e3f19d797cd79971da6c7/pkg/netutil/netutil_windows.go#L29) for the subnet unless the user explicitly passes in a different one.

> assuming the network doesn't already exists

Also correct, nerdctl does **zero** checking/creation of OS-side resources when the user creates a network (on Windows and Linux too afaik), all a `nerdctl network create` does is [write the CNI config JSON and call it a day](https://github.com/containerd/nerdctl/blob/4e3fd3b48f60fbce3b8e3f19d797cd79971da6c7/pkg/netutil/netutil.go#L265).

> An alternative to this (which I thought I mentioned elsewhere, I think there's a discussion about CNI network name/type behaviour in another ticket) would be to actually complete the TODO in this repo and remove network-creation support

IMHO, it should be expected, and even downright warranted of the CNI plugins to create underlying OS resources as needed (the [trivial reference Linux CNI plugins do this](https://github.com/containernetworking/plugins/blob/f95505231a8209b1b41eac2d5c2399c6ab444785/plugins/main/bridge/bridge.go#L336-L378), at least), so from my PoV the real question is what the WinCNI plugins should do in case of the HNS networks pre-existing with different properties, for which I have opened the following issue for further discussion: https://github.com/microsoft/windows-container-networking/issues/98

> I think there's a discussion about CNI network name/type behaviour in another ticket

I hope to contribute a patch which should alleviate the `BinaryName != HcnType` issue (https://github.com/microsoft/windows-container-networking/issues/57) which would hopefully not break any of the usecases these plugins are already used for.

An honest thank you for all your help in all of this, @TBBle!

@debj1t looks like the updated test suite has passed (at least for [0.2.0 configs, which are the only ones we currently test](https://github.com/microsoft/windows-container-networking/blob/d502b1b6d7e9e59f744dbc4903c3fa986668e7a4/test/utilities/connectivity_testing.go#L21)) just fine on this PR so please consider merging this!

**Comment by TBBle** [comment]
To move to using the 1.0.0 structures internally? Replace the use of "github.com/containernetworking/cni/pkg/types/040" with "github.com/containernetworking/cni/pkg/types/100" and adapt to any differences in the structures in code in this repo that interacts with them, as far as I know.

And yeah, nerdctl's code on Windows [starts with a blank IPAM structure](https://github.com/containerd/nerdctl/blob/v1.6.2/pkg/netutil/cni_plugin_windows.go#L47-L49), and [never populates the type field](https://github.com/containerd/nerdctl/blob/v1.6.2/pkg/netutil/netutil_windows.go#L86-L90). The comparable code on Unix has [different methods](https://github.com/containerd/nerdctl/blob/v1.6.2/pkg/netutil/cni_plugin_unix.go#L130-L134) for [different IPAM types](https://github.com/containerd/nerdctl/blob/v1.6.2/pkg/netutil/cni_plugin_unix.go#L142-L146) which prepopulate the type field.

I suspect the Windows-side nerdctl CNI generation implementation has been bitrotting a little, since it still references CNI 0.2.0, while the Unix side references 1.0.1 and 1.1.0.

**Comment by aznashwan** [comment]
Had already half-baked the reply to question 1 so posting it for posterity anyway:

> What is the work required to move to 1.0.0?

If you meant moving the internal structures used by the plugins to 1.0.0, it's a matter of switching the `containernetworking/cni/pkg/types/040` deps to `types/100`, which [**seemed** as trivial as dropping some fields](https://github.com/microsoft/windows-container-networking/pull/90/files#diff-8a47859b51210f40efd508827dea21a220fc7c0d7fee3f8d2a87ec38220a26baL378-L410) in my other closed PR, though I did not get to thoroughly test those changes since the testing framework was pretty faulty.

> I suspect the Windows-side nerdctl CNI generation implementation has been bitrotting a little

Indeed, the `nerdctl network create` code is a whole other can of worms, but it **should** be a completely separate issue from whether the plugins can handle 1.0.0 configs or not, and should hopefully not imply express updates to the plugins to facilitate it further.

File: cni/cni.go [hunk 1]
```diff
@@ -26,8 +26,8 @@ const (
 	Internal = "internal"
 )
 
-// Supported CNI versions.
-var VersionsSupported = []string{"0.2.0", "0.3.0"}
+// Oldest Supported CNI versions
+var OldestVersionSupported = "0.2.0"
```
Comment (iankingori) [inline cni/cni.go:30]: @Tbbie could we revert this to explicitly indicating supported versions? The current release versioning seems to correspond with greatest supported version. We can pull this change into a separate PR and discuss it.

Comment (TBBle) [reply cni/cni.go:30]: Just to be clear, you're requesting I open-code this list to the current return value of `cniVers.VersionsStartingFrom(OldestVersionSupported)`?

Comment (iankingori) [reply cni/cni.go:30]: That was the idea but seems counter productive. @Debjit would we be okay with using the [cni ](github.com/containernetworking/cni) import versioning to limit the max supported version? looking at the [releases ](https://github.com/containernetworking/cni/releases) it doesn't seem to be revved that frequently

Comment (TBBle) [reply cni/cni.go:30]: So I was poking around the CNI spec a little bit, and it's possible this change is faulty, as if I'm skimming it correctly, [by declaring we support config files 0.4.0 or newer, we are also opening ourselves up to having `CHECK` called for such config files](https://github.com/containernetworking/cni/blob/v1.1.2/pkg/skel/skel.go#L220-L241), but [we currently pass `nil` as the check callback](https://github.com/microsoft/windows-container-networking/blob/master/cni/plugin.go#L47-L48). A quick look at the CNI skeleton code [doesn't show any checks for `nil` in the relevant code](https://github.com/containernetworking/cni/blob/v1.1.2/pkg/skel/skel.go#L166-L172), so this would panic if we hit it. (`CHECK` was introduced in CNI spec 0.4.0 and it also changed the `DEL` API; 0.4.0 to 1.0.0 looks like it only changed config and structures, according to [the "major changes" list](https://github.com/containernetworking/cni/blob/main/SPEC.md#released-versions))

A quick look at BuildKit (my test case) doesn't show any calls to the go-cni `Check` API which wraps this, so I haven't exercised this change in my limited testing.

I am not totally on top of how CNI works, so when I was building this PR I was mostly looking at the config-file parser, but perhaps we need to (for purposes of this PR) keep the existing `VersionsSupported` array, and push the actual supported versions change out to require an implementation of `CHECK` as well. (And probably a more-careful check of changes in the spec)

If I'm right and the version bump was incorrect, my understanding is we can still use the latest library (we were already on library 0.8.1 which appears to have supported spec 0.4.0), so we only need to revert our version declaration to avoid the behaviour change. We can also separately go ahead and upgrade the internal structures to types/100 or later in the meantime, as again, we are already using version types/040 with spec 0.3.0 support. (Again, might be worth validating in the spec that the meaning of those fields hasn't changed, in case something changed that isn't caught by simple Go compiler type-checking.)

Comment (MikeZappa87) [reply cni/cni.go:30]: CHECK is not implemented in containerd or cri-o. I added the check code in go-cni however discontinued it as it involves a significant amount of work to the runtime/cri-api/kubelet.

Comment (TBBle) [reply cni/cni.go:30]: So maybe just a function that returns a "not implemented" error would be sufficient to avoid the panic, in case that rework is done in the future, or another CNI user appears.

Comment (iankingori) [reply cni/cni.go:30]: I see the [cni plugins](https://github.com/containernetworking/plugins/blob/abee8ccc0de94610e5d14488479bcb87ad64776c/plugins/main/windows/win-bridge/win-bridge_windows.go#L212C1-L215C2) callback function returns `nil` which emulates success? `not implemented` would allow us to catch unknown or future scenarios earlier though. @MikeZappa87 are we ok with adding a no-op check command callback + an issue to track the work to implement it?

Comment (debj1t) [reply cni/cni.go:30]: I think we can proceed with returning "not implemented" for CHECK, and adding both 0.4.0 and 1.0.0 as the supported versions in the array. We can track the remaining work with the below issues based on requirement to support them (doesn't seem to be widely used at the moment):
1. Implement CHECK; update config examples to include disableCheck field (0.4.0) 
2. Remove non-list configurations; remove version field of interfaces array (1.0.0)

**Comment by Lin-Huang-Wei** [comment]
> ```json
> nerdctlID
> ```
Hi  @TBBle 
I wonder how did you got the nerdctlID in Powershell ?

**Comment by aznashwan** [comment]
@Lin-Huang-Wei note that the `nerdctlID` field is only set by `nerdctl network create` when it is the one creating the network, but there's nothing stopping you from defining CNI networks yourself, in which case they will not have an ID.

Example containerd setup script creating ID-less CNI network:
https://github.com/containerd/containerd/blob/main/script/setup/install-cni-windows#L78-L97

**Comment by donno** [comment]
Thanks @TBBle, I was able to use this request/branch to get networking working with containerd.

I have someone else has since submitted a pull request, #101 which has merged. I'm yet to try building that and checking that but perhaps this require is due for being closed.

**Comment by TBBle** [comment]
Yeah, that makes sense, #101 appears to be the union of these changes, #90, and the resolution of the missing `CHECK` support discussion in this PR.

So I'll close this as redundant.
